### DISCUSSION TITLE
How can I set a list item's `key` based on the parent?

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2170] [Q&A] [answered]
Is there a way for a list item's key to be derived from a field of its parent?

I've read through the docs extensively regarding the normalized cache and resolvers. Since I'm still a bit new to urql/graphql, I think I may be missing how I can achieve this.

I'm fetching a user, and inside of that user, I'm fetching the artists they have liked.

Pseudo-code:

```graphql
query User {
  user(id: 123) {
    id
    likedArtists {
      slug
      name
      likedAt
    }
  }
}
```

The issue: **I want `LikedArtist`'s key to be the user ID + the artist slug**. The reason is that this like itself is relative to its parent user (such as the `likedAt` field.)

Currently, the key being set for each item is `User:123.likedArtists.0`. This is _almost_ what I want. I just want to replace the `0` at the end with the `slug` from a given liked artist.

Is this possible? I notice that the `cacheExchange`'s `keys` field doesn't give you access to the `parent` the way that `resolvers` does. Should I be doing this inside of `resolvers` somewhere?

One hack I considered: Add an `_id` field, and override it with the resolver. But this feels super hacky and wrong.

```graphql
query User {
  user(id: 123) {
    id
    likedArtists {
      slug
      name
      likedAt
      _id: slug # add this fake ID here
    }
  }
}
```

And then I could resolve the `_id` using the `parent`?

```ts
cacheExchange({
  resolvers: {
    LikedArtist: {
      _id(result, args, cache, info) {
        return removeTrailingIndex(info.parentKey) + result.slug
      }
    }
  }
})
```

The reason we don't have a unique ID for `likedArtist` is that this is constructed on the fly on the backend, and it doesn't exactly make sense to have a ID just for the frontend's normalized caching. I know it's considered bad practice to construct IDs with URQL, but I hope this use-case makes sense.

Thank you for your help.

**Comment by kitten** [comment, marked as answer]
No, to put it short it's not possible to alter an internal caching key to be based on a parent's key. This simply wouldn't make much sense in a relational system, because you're basically dictating anyway in that case that this entity is always related to its parent, meaning it's not globally unique but unique to a parent's key. In other words, the default of caching by embedded keys (which are internal and not meant to make sense outside of Graphcache's internal data structure) is sufficient to tie this item to its parent only.

So, the question is rather, what are you trying to achieve here? I'm almost positively certain that you may be looking for a workaround that'd be much better represented by "edge" types in your schema, but it's hard for me to tell.

> The reason we don't have a unique ID for likedArtist is that this is constructed on the fly on the backend

I can kind of see that you're likely already treating this as an edge without an edge type, which is a common "problem" I see in many schemas that are being made. This basically occurs when you're treating a given type, say here "Artist" as different because it's relational. Say we have a relation of `User -> Artist` but in this case you're changing the type to `LikedArtist` because of its relation. Instead of `Artist` and `LikedArtist` being the same now, you're now facing the problem of your schema basically saying that they're two distinct types; in other words, you've made an otherwise identical type slightly different to give it a semantic and implicit meaning.

Typically, the way to make this explicit is to make an edge type, e.g. `User -> LikedArtistEdge -> Artist`. Now, `LikedArtistEdge` is free to have a virtual ID or no ID at all and simply points to the entity it's referring to.

**Reply by nandorojo** [reply]
Thanks for the thorough response. I think I understand what you're getting at.

To clarify, are you suggesting we switch to this kind of structure?

```graphql
query User {
  user(id: 123) {
    id
    likedArtistEdges {
      id # virtual ID, made for the frontend
      likedAt
      node {
        slug
        name
      }
    }
  }
}
```

**Reply by kitten** [reply]
It depends of course; I'm not sure when this would come into play and only assumed that the problem is that your queries weren't allowing you to associate a presumed `Artist` entity with a `LikedArtist` 🤔 But assuming that's the case, yea, this should 100% help.
I'd also argue that it's a pretty natural schema design. If we consider metadata like `likedAt` and `name`, one clearly relates to an entity that isn't at all related to `User` while the other is describing their relation, i.e. the edge, so it feels natural to me to make an edge type for it.

**Reply by nandorojo** [reply]
Yeah, I think I should give a better example to clarify the complication here.

```graphql
query Investor {
  investor(id: 123) {
    id
    positions {
      symbol
      quantity
      marketValue
    }
  }
}
```

In this case, each position is uniquely identified by the combination of its `symbol` and the parent's `userId`. `Artist` was a bad example because that can globally exist as its own entity. A `Position`, on the other hand, must be associated to its parent `Investor`. The backend gets the `Investor`, and based on the investor's trade history + market prices, constructs the positions. The issue is, the keys for positions becomes `Investor:123.positions.0`. 

One solution is to add `investor_id` to the `position` and construct the key with the `cacheExchange`'s `keys`. But this feels like a NoSQL solution that shouldn't be necessary for GraphQL, especially given that our query already has the `Investor` inside of it. 

Would `investor_id` be better suited for an "edge type" as you've mentioned? I'm sure it's hard for you to give an opinion from the outside.

Curious to get your thoughts there. Really appreciate it!

> I'd also argue that it's a pretty natural schema design. If we consider metadata like likedAt and name, one clearly relates to an entity that isn't at all related to User while the other is describing their relation, i.e. the edge, so it feels natural to me to make an edge type for it.

Good to know.

**Reply by JoviDeCroock** [reply]
I think for this example it comes down to being an 'embedded' entity, where returning null as a key makes it embed into the parent which sounds like the intention here with Investor:id.pos

**Reply by nandorojo** [reply]
Right. I could still see a situation where the key-based caching is useful, though, such as on the position detail screen. For example, if I fetch an individual position by its symbol, I would expect this to pull from cached values in the list view.

```graphql
query Investor {
  investor(id: 123) {
    id
    position(symbol: "AAPL") {
      symbol
      quantity
      marketValue
    }
  }
}
```

**Reply by kitten** [reply]
Slight problem here, sorry 😅

> The issue is, the keys for positions becomes Investor:123.positions.0.

I got to be honest. I think this doesn't illustrate what issue you're having but it reiterates a prior point and does not actually explain the problem you're facing 

I don't see any problem here with the embedded key that is in conflict with what you're describing. I can see how the parent's key together with the sub-entity's symbol is unique, but I'd argue that that doesn't describe a conflict with the embedded nature of your data 

> For example, if I fetch an individual position by its symbol, I would expect this to pull from cached values in the list view.

I think this is still a good argument for a virtual key though in that case, maybe in addition to edges, if possible. If an entity is presented in multiple places as derived data but is globally unique then there's a precedent to say that it must be unique from a set of attributes that aren't exposed on the schema right now, e.g. either there's a backlink from a `Position` to an `Investor` (say with an `investorId` field) which would then have to be made part of `keys`) or a virtual key must exist to reflect that 

Basically, I mean, when `Position` is accessible from a query that doesn't necessitate it being in a single place of the schema, then it must have fields reflecting that, not for the sake of Graphcache only, but because it must fully represent what it's for 🤔
